### DISCUSSION TITLE
Support for single-file "collections"

### DISCUSSION THREAD

**Discussion by llibdude** [tinacms/tinacms#2948] [Help] [answered]
Title might not be very clear, so here is an example.

Say I want to have a blog, and for each blog you can assign it one more tags.  All a tag is is a name.  And lets say also we have tens or even one hundred tags.

To me it seems overkill to generate a markdown file per `tag`, but I also don't want to hard code a list in the schema itself as I want my content creators to be able to add/edit/remove tags.

Is my best/only option to still have a document per tag?

**Comment by jeffsee55** [comment, marked as answer]
What are you hoping to do with the tag? Why can't you do this?:

```ts
defineSchema({
  collections: [{
    name: "post",
    label: "Post",
    fields: [{
      name: 'tag',
      options: ['css', 'html', 'javascript']
      list: true
    }]
  }]
})
```

**Reply by llibdude** [reply]
I ultimately would like to filter on the tag, so I'm hoping the api allows for that.  But the reason I don't want to hard code it is so the content creators on my team can add/remove/edit tags without needing to go into the code.

**Reply by jeffsee55** [reply]
Ah ok, you would be able to filter on it, but adding new ones - yes would require it to be a document.

**Reply by jeffsee55** [reply]
We've talked about having a collection be backed by a single file (like `.csv` or `.json`) which is just an array of records, this feels like a good use case for that:

```ts
defineSchema({
  collections: [{
    name: "post",
    label: "Post",
    path: "content/posts",
    fields: [{
      name: 'tags',
      list: true,
      type: 'reference',
      collection: ['tag']
    }]
  }, {
    name: "tag",
    label: "Tag",
    path: "content/tags.csv",
    fields: [{
      name: 'name',
      type: 'string'
    }]
  }]
})
```

We don't have plans to actually do this at this time FYI

**Reply by llibdude** [reply]
Yeah I think that's what I would want for this case.  Thanks for the help!

**Reply by elhe26** [reply]
@llibdude did it work?

**Reply by jeffsee55** [reply]
@elhe26 this isn't actually supported, I was just trying to make sure I understood the need. Currently you'll need a new document for each item.

**Reply by coreyaus** [reply]
Love that idea Jeff and it'd provide an elegant solution to a few things like tags, theme colours, etc. **Is there a public GitHub issue for the idea you outlined where we can subscribe to follow any future discussion and updates?**

-----

For what it's worth, here's how we're currently handling this to avoid creating a document for each tag:

- We've made tags and similar items editable via a catch-all "Theme" collection (where tags, colours, etc. all have their own templates)
    - We have a single file for each of these items (i.e. `tags.json`, `colours.json`, etc.)
    - The templates ensure each file contains an array of items (i.e. an array of tags in the case of `tags.json`, where editors can create new tags and update existing ones)
- We then import the relevant `.json` file to build the options for our tag fields (example below):
    - This works really well, though obviously there are some downsides - e.g. there can be a slight delay before seeing new tag options in the tag dropdown fields in production

I'm not fully sold on this approach, so think what Jeff outlines sounds promising - but for now I'm just happy to avoid having separate documents for each tag and other similar utilities.

```typescript
import data from '../../content/theme/tags.json'
import type { TinaField } from '@tinacms/cli'

const allTags = data.tags?.map((tag) => ({
  label: tag.name,
  value: tag.name,
}))

const tagField: TinaField = {
  type: 'string',
  name: 'tags',
  label: 'Tags',
  options: allTags,
  list: true
};
```

**Reply by jeffsee55** [reply]
Very clever 😆 

I've renamed this discussion to reflect the idea of single-file collections. We'll update when we start work on it.
